File: SASMopubRewardedVideoCustomEvent.m
//
//  SASMopubRewardedVideoCustomEvent.m
//  SmartAdServer
//
//  Created by Thomas Geley on 22/12/2016.
//  Copyright © 2019 Smart AdServer. All rights reserved.
//

#import "SASMopubRewardedVideoCustomEvent.h"
#import "MPRewardedVideoError.h"
#import "MPReward.h"
#import "MPLogging.h"
#import "SASMopubUtils.h"
#import <SASDisplayKit/SASDisplayKit.h>

NS_ASSUME_NONNULL_BEGIN

@interface SASMopubRewardedVideoCustomEvent () <SASRewardedVideoManagerDelegate>

@property (nonatomic, strong, nullable) SASRewardedVideoManager *rewardedVideoManager;

@end

@implementation SASMopubRewardedVideoCustomEvent
@dynamic delegate;
@dynamic localExtras;
@dynamic hasAdAvailable;
@dynamic isRewardExpected;

#pragma mark - Object Lifecycle

- (void)dealloc {
    [self destroyRewardedVideoManager];
}

- (void)destroyRewardedVideoManager {
    self.rewardedVideoManager.delegate = nil;
    self.rewardedVideoManager = nil;
}

#pragma mark - MPFullscreenAdAdapter Override

- (BOOL)isRewardExpected {
    return YES;
}

- (BOOL)hasAdAvailable {
    return self.rewardedVideoManager.adStatus == SASAdStatusReady;
}

- (BOOL)enableAutomaticImpressionAndClickTracking {
    return YES;
}

- (void)requestAdWithAdapterInfo:(NSDictionary *)info adMarkup:(NSString * _Nullable)adMarkup {

    // Reset rewarded video manager
    [self destroyRewardedVideoManager];
    
    // Extracting placement from custom event info
    NSError *error = nil;
    SASAdPlacement *adPlacement = [SASMopubUtils adPlacementWithCustomEventInfo:info error:&error];
    
    if (adPlacement == nil) {
        // Failing if custom info are invalid
        [self.delegate fullscreenAdAdapter:self didFailToLoadAdWithError:error];
        return;
    }
    
    // Creating a new rewarded video manager
    self.rewardedVideoManager = [[SASRewardedVideoManager alloc] initWithPlacement:adPlacement delegate:self];
    
    // Loading ad from ad placement
    [self.rewardedVideoManager load];
    
}

- (void)presentAdFromViewController:(UIViewController *)viewController {
    if (self.rewardedVideoManager.adStatus == SASAdStatusReady) {
        [self.rewardedVideoManager showFromViewController:viewController];
    }
}

- (void)handleDidPlayAd {
    if (self.rewardedVideoManager.adStatus == SASAdStatusExpired) {
        [self.delegate fullscreenAdAdapterDidExpire:self];
    }
}

#pragma mark - Rewarded video manager

- (void)rewardedVideoManager:(SASRewardedVideoManager *)manager didLoadAd:(SASAd *)ad {
    MPLogInfo(@"Smart Rewarded Video did load");
    [self.delegate fullscreenAdAdapterDidLoadAd:self];
}

- (void)rewardedVideoManager:(SASRewardedVideoManager *)manager didFailToLoadWithError:(NSError *)error {
    MPLogInfo(@"Smart Rewarded Video did fail to load with error: %@", error.localizedDescription);
    [self.delegate fullscreenAdAdapter:self didFailToLoadAdWithError:error];
}

- (void)rewardedVideoManager:(SASRewardedVideoManager *)manager didFailToShowWithError:(NSError *)error {
    MPLogInfo(@"Smart Rewarded Video did fail to show with error: %@", error.localizedDescription);
    [self.delegate fullscreenAdAdapter:self didFailToLoadAdWithError:error];
}

- (void)rewardedVideoManager:(SASRewardedVideoManager *)manager didAppearFromViewController:(UIViewController *)viewController {
    MPLogInfo(@"Smart Rewarded Video did appear");
    [self.delegate fullscreenAdAdapterAdWillAppear:self];
    [self.delegate fullscreenAdAdapterAdDidAppear:self];
}

- (void)rewardedVideoManager:(SASRewardedVideoManager *)manager didDisappearFromViewController:(UIViewController *)viewController {
    MPLogInfo(@"Smart Rewarded Video did disappear");
    [self.delegate fullscreenAdAdapterAdWillDisappear:self];
    [self.delegate fullscreenAdAdapterAdDidDisappear:self];
}

- (void)rewardedVideoManager:(SASRewardedVideoManager *)manager didCollectReward:(SASReward *)reward {
    MPLogInfo(@"Smart Rewarded Video did collect reward");
    MPReward *moPubReward = [[MPReward alloc] initWithCurrencyType:reward.currency amount:reward.amount];
    [self.delegate fullscreenAdAdapter:self willRewardUser:moPubReward];

}

- (BOOL)rewardedVideoManager:(SASRewardedVideoManager *)manager shouldHandleURL:(NSURL *)URL {
    MPLogInfo(@"Smart Rewarded Video did receive tap");
    [self.delegate fullscreenAdAdapterDidReceiveTap:self];
    return YES;
}

@end

NS_ASSUME_NONNULL_END

File: SASMopubRewardedVideoCustomEvent.swift
//
//  SASMopubRewardedVideoCustomEvent.swift
//  SASMopubRewardedVideoCustomEvent
//
//  Created by Loïc GIRON DIT METAZ on 01/09/2021.
//

import Foundation
import SASDisplayKit
import MoPubSDK

@objc(SASMopubRewardedVideoCustomEvent)
class SASMopubRewardedVideoCustomEvent: MPFullscreenAdAdapter {
    
    private var rewardedVideoManager: SASRewardedVideoManager? = nil
    
    override func requestAd(withAdapterInfo info: [AnyHashable : Any], adMarkup: String?) {
        do {
            // Extracting placement from custom event info
            let adPlacement = try SASMopubUtils.adPlacement(customEventInfo: info)
            
            // Creating a new interstitial manager
            rewardedVideoManager = SASRewardedVideoManager(placement: adPlacement, delegate: self)
            
            // Loading ad from ad placement
            rewardedVideoManager?.load()
        } catch {
            // Failing if custom info are invalid
            delegate?.fullscreenAdAdapter(self, didFailToLoadAdWithError: error)
        }
    }
    
    override func presentAd(from viewController: UIViewController) {
        if rewardedVideoManager?.adStatus == .ready {
            rewardedVideoManager?.show(from: viewController)
        }
    }
    
    override var enableAutomaticImpressionAndClickTracking: Bool {
        return true
    }
    
}

extension SASMopubRewardedVideoCustomEvent: SASRewardedVideoManagerDelegate {
    
    func rewardedVideoManager(_ manager: SASRewardedVideoManager, didLoad ad: SASAd) {
        delegate?.fullscreenAdAdapterDidLoadAd(self)
    }
    
    func rewardedVideoManager(_ manager: SASRewardedVideoManager, didFailToLoadWithError error: Error) {
        delegate?.fullscreenAdAdapter(self, didFailToLoadAdWithError: error)
    }
    
    func rewardedVideoManager(_ manager: SASRewardedVideoManager, didFailToShowWithError error: Error) {
        delegate?.fullscreenAdAdapter(self, didFailToShowAdWithError: error)
    }
    
    func rewardedVideoManager(_ manager: SASRewardedVideoManager, didAppearFrom viewController: UIViewController) {
        delegate?.fullscreenAdAdapterAdWillAppear(self)
        delegate?.fullscreenAdAdapterAdDidAppear(self)
    }
    
    func rewardedVideoManager(_ manager: SASRewardedVideoManager, didDisappearFrom viewController: UIViewController) {
        delegate?.fullscreenAdAdapterAdWillDisappear(self)
        delegate?.fullscreenAdAdapterAdDidDisappear(self)
    }
    
    func rewardedVideoManager(_ manager: SASRewardedVideoManager, didCollect reward: SASReward) {
        if let reward = MPReward(currencyType: reward.currency, amount: reward.amount) {
            delegate?.fullscreenAdAdapter(self, willRewardUser: reward)
        }
    }
    
    func rewardedVideoManager(_ manager: SASRewardedVideoManager, didClickWith URL: URL) {
        delegate?.fullscreenAdAdapterDidReceiveTap(self)
    }
    
}
